ckanext/opendata_theme/opengov_custom_header/controller.py

```diff
@@ -0,0 +1,184 @@
+            self.redirect_to_custom_header_page(header_data)
+
+    def add_link(self):
+        if request.method == 'POST':
+            header_data = self.get_custom_header_metadata()
+            data = clean_dict(dict_fns.unflatten(
+                tuplize_dict(parse_params(request.POST))))
+            header_data['links'].append(
+                Header(
+                    title=data.get('new_title'),
+                    link=data.get('new_link'),
+                    position=len(header_data['links'])+1,
+                ))
+            error = self.save_header_metadata(header_data)
+            header_data['errors'] = error
+            return render('admin/custom_header.html',
+                          extra_vars=header_data)
+
+    def custom_header(self):
+        custom_header = self.get_custom_header_metadata()
+        if not custom_header.get('links'):
+            # this block is required for base initialization
+            # it happens only once when default custom header metadata is not set
+            # because it is set in build_pages_nav_main helper function which is called
+            # during the page rendering.
+            # reset_custom_header is able to render the page in the background for setting default metadata.
+            self.reset_custom_header()
+        if request.method == 'POST':
+            data = clean_dict(dict_fns.unflatten(
+                tuplize_dict(parse_params(request.POST))))
+            custom_header = {
+                'links': [],
+                'layout_type': data.get('layout_type', 'default')
+            }
+            data_len = len(data.get('link', []))
+            if data_len > 1:
+                for index in range(data_len):
+                    custom_header['links'].append(Header(
+                        title=data['title'][index],
+                        link=data['link'][index],
+                        position=data['position'][index],
+                    ))
+            else:
+                custom_header['links'].append(Header(
+                    title=data['title'],
+                    link=data['link'],
+                    position=data['position'],
+                ))
+            error = self.save_header_metadata(custom_header)
+            custom_header['errors'] = error
+
+        return render('admin/custom_header.html',
+                      extra_vars=custom_header)
+
+    def redirect_to_custom_header_page(self, vars):
+        redirect_to(
+            controller=CONTROLLER,
+            action='custom_header',
+            extra_vars=vars
+        )
+
+    def reset_custom_header(self):
+        custom_header = {
+            'layout_type': 'default',
+        }
+        self.save_default_header_metadata(custom_header)
+        self.save_header_metadata(custom_header)
+        # for setting default headers from page to db
+        render('admin/custom_header.html',
+               extra_vars=custom_header)
```

Comment:
It's not necessary to render the template since we want to redirect users.

Reply:
you are right)

Reply:
but this rendering is a point where the infinite loop mentioned later is broken. because during the rendering we setting default header metadata for right rendering later with controller method custom_header.
the issue here is more with page initialization with a controller, because we need default header metadata for regular page rendering, but for getting default header metadata we have to render the page just because we can set them only during the page rendering.